bootstrap/src/stream.ml

```diff
@@ -0,0 +1,576 @@
+open Rudiments
+
+type 'a elm =
+  | Nil
+  | Cons of 'a * 'a t
+and 'a t = 'a elm lazy_t
+
+let empty = lazy Nil
+
+let init n ~f =
+  let rec fn i n f = lazy begin
+    match i < n with
+    | false -> Nil
+    | true -> begin
+      let elm = f i in
+      let t = fn (Usize.succ i) n f in
+      Cons(elm, t)
+    end
+  end in
+  fn 0 n f
+
+let length t =
+  let rec fn t i = begin
+    match t with
+    | (lazy Nil) -> i
+    | (lazy (Cons(_, t'))) -> fn t' (Usize.succ i)
+  end in
+  fn t 0
+
+let is_empty = function
+  | lazy Nil -> true
+  | lazy (Cons(_, _)) -> false
+
+let hd = function
+  | lazy Nil -> halt "Empty stream has no head"
+  | lazy (Cons(elm, _)) -> elm
+
+let tl = function
+  | lazy Nil -> halt "Empty stream has no tail"
+  | lazy (Cons(_, t)) -> t
+
+let push' elm s =
+  Cons(elm, s)
+
+let push elm s =
+  lazy (push' elm s)
+
+let pop = function
+  | lazy Nil -> halt "Empty stream has no head"
+  | lazy (Cons(elm, t)) -> (elm, t)
+
+let rec concat t0 t1 = lazy begin
+  match t0 with
+  | (lazy Nil) -> Lazy.force t1
+  | (lazy (Cons(elm, t0'))) -> push' elm (concat t0' t1)
+end
+
+let rec take t n =
+  lazy begin
+    match t, n with
+    | _, 0 -> Nil
+    | (lazy Nil), _ -> Lazy.force t
+    | (lazy (Cons(elm, t'))), _ -> push' elm (take t' (n - 1))
```

Comment:
`(n - 1)` can be written as `(pred n)`.  I don't have a strong preference on which to use, but it's probably best to be consistent about using `succ/pred` vs `+- 1`